cmd/kops/replace.go

```diff
@@ -84,138 +99,314 @@ func NewCmdReplace(f *util.Factory, out io.Writer) *cobra.Command {
 }
 
 // RunReplace processes the replace command
-func RunReplace(ctx context.Context, f *util.Factory, out io.Writer, c *ReplaceOptions) error {
+func RunReplace(ctx context.Context, f *util.Factory, out io.Writer, opt *ReplaceOptions) error {
 	clientset, err := f.KopsClient()
 	if err != nil {
 		return err
 	}
 
-	for _, f := range c.Filenames {
+	var objects []runtime.Object
+	for _, filename := range opt.Filenames {
 		var contents []byte
-		if f == "-" {
+		if filename == "-" {
 			contents, err = ConsumeStdin()
 			if err != nil {
 				return err
 			}
 		} else {
-			contents, err = vfs.Context.ReadFile(f)
+			contents, err = vfs.Context.ReadFile(filename)
 			if err != nil {
-				return fmt.Errorf("error reading file %q: %v", f, err)
+				return fmt.Errorf("error reading file %q: %v", filename, err)
 			}
 		}
 		sections := text.SplitContentToSections(contents)
 
 		for _, section := range sections {
-			o, gvk, err := kopscodecs.Decode(section, nil)
+			o, _, err := kopscodecs.Decode(section, nil)
 			if err != nil {
-				return fmt.Errorf("error parsing file %q: %v", f, err)
+				return fmt.Errorf("error parsing file %q: %v", filename, err)
 			}
 
-			switch v := o.(type) {
-			case *kopsapi.Cluster:
-				{
-					// Retrieve the current status of the cluster.  This will eventually be part of the cluster object.
-					cloud, err := cloudup.BuildCloud(v)
-					if err != nil {
-						return err
-					}
-					status, err := cloud.FindClusterStatus(v)
-					if err != nil {
-						return err
-					}
+			objects = append(objects, o)
+		}
+	}
 
-					// Check if the cluster exists already
-					clusterName := v.Name
-					cluster, err := clientset.GetCluster(ctx, clusterName)
-					if err != nil {
-						if errors.IsNotFound(err) {
-							cluster = nil
-						} else {
-							return fmt.Errorf("error fetching cluster %q: %v", clusterName, err)
-						}
-					}
-					if cluster == nil {
-						if !c.Force {
-							return fmt.Errorf("cluster %v does not exist (try adding --force flag)", clusterName)
-						}
-						_, err = clientset.CreateCluster(ctx, v)
-						if err != nil {
-							return fmt.Errorf("error creating cluster: %v", err)
-						}
-					} else {
-						_, err = clientset.UpdateCluster(ctx, v, status)
-						if err != nil {
-							return fmt.Errorf("error replacing cluster: %v", err)
-						}
-					}
+	for _, obj := range objects {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+
+		clusterName := ""
+
+		switch v := obj.(type) {
+		case *kopsapi.Cluster:
+			clusterName = v.Name
+		case *kopsapi.InstanceGroup:
+			clusterName = v.ObjectMeta.Labels[kopsapi.LabelClusterName]
+		case *kopsapi.SSHCredential:
+			clusterName = v.ObjectMeta.Labels[kopsapi.LabelClusterName]
```

Comment:
Existing behavior, but this seems to me to be an abuse of a label. We're treating it sort-of like `namespace`, but we're shoehorning it into a magic label because `ObjectMeta` doesn't have a `Cluster` field.

I think we need a better way of handling cluster-namespaced objects. Perhaps by putting the cluster name in the `Namespace` field?